src/main/flight/imu.c

```diff
@@ -478,10 +496,29 @@ static void imuCalculateEstimatedAttitude(timeUs_t currentTimeUs)
     }
 #endif
 #if defined(USE_GPS)
-    if (!useMag && STATE(FIXED_WING) && sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat >= 5 && gpsSol.groundSpeed >= 300) {
-        // In case of a fixed-wing aircraft we can use GPS course over ground to correct heading
-        rawYawError = DECIDEGREES_TO_RADIANS(attitude.values.yaw - gpsSol.groundCourse);
-        useYaw = true;
+    if (!useMag && sensors(SENSOR_GPS) && STATE(GPS_FIX) && gpsSol.numSat >= 5 && gpsSol.groundSpeed >= GPS_COG_MIN_GROUNDSPEED) {
+        // Use GPS course over ground to correct attitude.values.yaw
+        if (STATE(FIXED_WING)) {
+            courseOverGround = DECIDEGREES_TO_RADIANS(gpsSol.groundCourse);
+            useCOG = true;
+        } else {
+            // If GPS rescue mode is active and we can use it, go for it.  When we're close to home we will 
+            // probably stop re calculating GPS heading data.  Other future modes can also use this extern
+
+            if (canUseGPSHeading) {
+                courseOverGround = DECIDEGREES_TO_RADIANS(gpsSol.groundCourse);
+
+                useCOG = true;
+            }
+        }
+
+        if (useCOG && shouldInitializeGPSHeading()) {
+            // Reset our reference and reinitialize quaternion.  This will likely ideally happen more than once per flight, but for now,
+            // shouldInitializeGPSHeading() returns true only once.
+            imuComputeQuaternionFromRPY(&qP, attitude.values.roll, attitude.values.pitch, gpsSol.groundCourse);
```

Comment:
This is IMO far too fragile. Either wait a bit until heading converges (time constant should be few seconds), or calculate error and update orientation quaternion in one step (rotating it around Z axis).